Sources/OpenCombine/Publishers/Publishers.ReplaceError.swift

```diff
@@ -0,0 +1,134 @@
+//
+//  Publishers.ReplaceError.swift
+//  OpenCombine
+//
+//  Created by Bogdan Vlad on 8/29/19.
+//
+
+extension Publishers {
+    /// A publisher that replaces any errors in the stream with a provided element.
+    public struct ReplaceError<Upstream>: Publisher where Upstream: Publisher {
+
+        /// The kind of values published by this publisher.
+        public typealias Output = Upstream.Output
+
+        /// The kind of errors this publisher might publish.
+        ///
+        /// Use `Never` if this `Publisher` does not publish errors.
+        public typealias Failure = Never
+
+        /// The element with which to replace errors from the upstream publisher.
+        public let output: Upstream.Output
+
+        /// The publisher from which this publisher receives elements.
+        public let upstream: Upstream
+
+        public init(upstream: Upstream,
+                    output: Publishers.ReplaceError<Upstream>.Output) {
+            self.upstream = upstream
+            self.output = output
+        }
+
+        /// This function is called to attach the specified `Subscriber`
+        /// to this `Publisher` by `subscribe(_:)`
+        ///
+        /// - SeeAlso: `subscribe(_:)`
+        /// - Parameters:
+        ///     - subscriber: The subscriber to attach to this `Publisher`.
+        ///                   once attached it can begin to receive values.
+        public func receive<SubscriberType: Subscriber>(subscriber: SubscriberType)
+            where Upstream.Output == SubscriberType.Input,
+                  SubscriberType.Failure == Failure
+        {
+            let replaceErrorSubscriber = _ReplaceError<Upstream, SubscriberType>(
+                downstream: subscriber,
+                output: output
+            )
+            upstream.subscribe(replaceErrorSubscriber)
+        }
+    }
+}
+
+extension Publisher {
+    /// Replaces any errors in the stream with the provided element.
+    ///
+    /// If the upstream publisher fails with an error, this publisher emits the provided
+    /// element, then finishes normally.
+    /// - Parameter output: An element to emit when the upstream publisher fails.
+    /// - Returns: A publisher that replaces an error from the upstream publisher with
+    ///            the provided output element.
+    public func replaceError(with output: Self.Output) -> Publishers.ReplaceError<Self> {
+        return Publishers.ReplaceError(upstream: self, output: output)
+    }
+}
+
+private final class _ReplaceError<Upstream: Publisher, Downstream: Subscriber>
+    : OperatorSubscription<Downstream>,
+      Subscriber,
+      CustomStringConvertible,
+      Subscription
+    where Upstream.Output == Downstream.Input
+{
+    typealias Input = Upstream.Output
+    typealias Failure = Upstream.Failure
+
+    var hasAnyDownstreamDemand: Bool = false
```

Comment:
I'm not sure whether having a flag is enough.

When the `receive(_ input: Input)` method is called, downstream demand actually decreases and may reach 0, but this flag is not reset in that case.

Can you add a test fo the following scenario?

1. Downstream initially requests `.max(1)`.
2. Upstream sends one value, downstream receives it and returns zero demand.
3. Upstream sends failure.
4. Downstream should not receive anything at this point.
5. Downstream requests some non-zero demand.
6. And now it receives the replacement value.

If this test fails, we'll probably need to replace this flag with a demand counter, which will be decremented each time `receive(_ input: Input)` method is called.

Reply:
Yes, you are right. Good catch.